src/utils/checkIfFileValid.js

```diff
@@ -0,0 +1,25 @@
+// this function checks the validity of the file
+// this is done in the data-management module - upload samples
+import mime from 'mime-types';
+import path from 'path';
+
+import techOptions from './fileUploadSpecifications';
+
+const checkIfFileValid = (fileName, selectedTech) => {
+  const isValidType = (
+    techOptions[selectedTech].validMimeTypes
+      .includes(
+        mime.lookup(fileName),
+      )
+    || techOptions[selectedTech].validExtensionTypes
+      .includes(
+        path.extname(fileName),
+      )
+  );
+  const acceptedFilesRegexp = `(${techOptions[selectedTech].acceptedFiles.join('|')})$`;
```

Comment:
I know that naively writing unions using `join('|')` can cause issues with escaping. Might not be an issue here but Ruby has `Regexp.union` for this, I wonder if we could do something similar.

Reply:
what do you mean i didnt get it - .join just adds the character between every element in the array. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join Or you mean something else?

Reply:
What I mean is that the things you are joining are strings being turned into a regex. However the strings (file names) contain dots, which have a special meaning in regex, so technically need to be escaped. Sorry, I know it's weird to use another language as an example but you will see that `Regexp.union` safely escapes special characters:

```irb
irb(main):012:0> Regexp.union(*acceptedFiles)
=> /barcodes\.tsv|barcodes\.tsv\.gz|features\.tsv|features\.tsv\.gz|genes\.tsv|genes\.tsv\.gz|matrix\.mtx|matrix\.mtx\.gz/
irb(main):013:0> acceptedFiles.join('|')
=> "barcodes.tsv|barcodes.tsv.gz|features.tsv|features.tsv.gz|genes.tsv|genes.tsv.gz|matrix.mtx|matrix.mtx.gz"
```
Because dots mean "match any one character", my prediction is that the current logic will allow any characters in place of the dots. This is easy to test.

Also, I know you did not write this and just moved it, but I just noticed it here!

Reply:
OK, I think the mime type check prevents this from ever actually being an issue but it is technically wrong.

Reply:
We could use https://stackoverflow.com/a/29700268/7799676.

```js
const acceptedFileRegexp =  techOptions[selectedTech].acceptedFiles.map(_.escapeRegExp).join('|')
```

Seems to be correct

Reply:
![image](https://user-images.githubusercontent.com/68030850/121537354-c3918400-c9fb-11eb-82f3-ebb299c9f595.png)
Yeah, it is actually possible to witness this on the platform (the upload fails but beforehand it is marked as valid). My fix should work I think.

Reply:
Thanks for the explanation! I changed it.